nwbinspector/register_checks.py

```diff
@@ -0,0 +1,97 @@
+"""Authors: Cody Baker and Ben Dichter."""
+from collections import defaultdict, OrderedDict
+from functools import wraps
+
+import h5py
+
+global available_checks
+CRITICAL_IMPORTANCE = 2
+BEST_PRACTICE_VIOLATION = 1
+BEST_PRACTICE_SUGGESTION = 0
+
+importance_levels = OrderedDict(
+    CRITICAL_IMPORTANCE=CRITICAL_IMPORTANCE,
+    BEST_PRACTICE_VIOLATION=BEST_PRACTICE_VIOLATION,
+    BEST_PRACTICE_SUGGESTION=BEST_PRACTICE_SUGGESTION,
+)
+levels_to_importance = {v: k for k, v in importance_levels.items()}
+available_checks = OrderedDict({importance_level: defaultdict(list) for importance_level in importance_levels})
+
+# For strictly internal use only
+HIGH_SEVERITY = 5
+LOW_SEVERITY = 4
+severity_levels = OrderedDict({"HIGH_SEVERITY": HIGH_SEVERITY, "LOW_SEVERITY": LOW_SEVERITY, None: 3})
+levels_to_severity = {v: k for k, v in severity_levels.items()}
+
+
+def register_check(importance, neurodata_type):
+    """Wrap a check function to add it to the list of default checks for that severity and neurodata type."""
+
+    def register_check_and_auto_parse(check_function):
+        if importance not in importance_levels.values():
+            raise ValueError(
+                f"Indicated importance ({importance}) of custom check ({check_function.__name__}) is not a valid "
+                "importance level! Please choose from [CRITICAL_IMPORTANCE, BEST_PRACTICE_VIOLATION, "
+                "BEST_PRACTICE_SUGGESTION]."
+            )
+        check_function.importance = levels_to_importance[importance]
+        check_function.neurodata_type = neurodata_type
+
+        @wraps(check_function)
+        def auto_parse_some_output(*args, **kwargs):
+            if args:
+                obj = args[0]
+            else:
+                obj = kwargs[list(kwargs)[0]]
+            auto_parsed_result = check_function(*args, **kwargs)
+            if auto_parsed_result is not None:
+                auto_parsed_result.update(
+                    importance=check_function.importance,
+                    severity=levels_to_severity[auto_parsed_result.get("severity", severity_levels[None])],
+                    check_function_name=check_function.__name__,
+                    object_type=type(obj).__name__,
+                    object_name=obj.name,
+                    location=parse_location(nwbfile_object=obj),
+                )
```

Comment:
I would suggest to create a data class to represent inspector messages, rather than hand-written dicts. Here an example of how this class could look like. 

```
from dataclasses import dataclass

@dataclass
class InspectorMessage:
   # User-defined parameters
   severity: Severity
   message: str

   # Parameters that are usually set py the register_class decorator
   importance: Importance = Importance.BEST_PRACTICE_SUGGESTION
   check_function_name: str = ""
   object_type: str = ""
   object_name: str = ""
   location: str = None
   
   def update(self,
              importance: Importance,
              check_function_name: str,
              object_type: str, 
              object_name: str,
              location: str):
      """Set parameter values used internally by the NWBInspector"""
      self.importance = importance
      self.check_function_name = check_function_name
      self.object_type = object_type
      self.object_name = object_name
      self.location = location
```

By defining a data class we can define the API more rigorously because we can explicitly define which parameter we expect for the return messages and we can also use type annotations to explictly define what kind of data we expect for each parameter. E.g., by defining the importance levels as an Enum type (see my above suggestion), we can define that the importance parameter should use the values defined in the ``Importance`` enumeration. Another issue this avoids, is that when using hand-written dicts a user may write a new check and accidentally omit or misspell a required key. With hand-written dicts, this will result in an error at runtime somewhere in the internals of the nwbinspector because we are trying to access a key that doesn't exists. In contrast, when we define this as a class, the code will fail already during parsing of the code because Python will raise an error that a parameter is missing or unknown. When using the ``@dataclass`` approach we also get a nice default representation when we want to print ``InspectorMessage`` objects.

Normally I would use ``NamedTuple`` from the ``typing`` module as base class for the ``InsepctorMessage`` class, but  ``NamedTuple`` is immutable and since we here want to be able to update the values of the ``object_name`` etc. in the decorator,  a mutable ``dataclass`` seems simpler and should give us all the functionality we need. 

I added an ``update`` method to the class here, mainly so that the register_check function doesn't need to change and to encapsulate the internal updates in a single function call, but I don't think this is required.

Reply:
I was a bit hesitant at first since this seemed like it was doing almost the same thing, but see https://github.com/catalystneuro/nwbinspector/pull/10 where I do conclude this can help make this safer in a number of ways.